File: codejam/python/2009/1B/1B_DecisionTree.py
import sys
from collections import defaultdict, deque, namedtuple

## Input crap
infile = sys.stdin
intokens = deque()
def getTokens(): 
    while not intokens:
        for c in infile.readline().rstrip().split() :
            intokens.append(c)    
def gs(): getTokens(); return intokens.popleft()
def gi(): return int(gs())
def gf(): return float(gs())
def gbs(): return [c for c in gs()]
def gis(n): return [gi() for i in range(n)]
def ia(m): return [0] * m
def iai(m,v): return [v] * m
def twodi(n,m,v): return [iai(m,v) for i in range(n)]
def fill2(m) : r = gis(2*m); return r[0::2],r[1::2]
def fill3(m) : r = gis(3*m); return r[0::3],r[1::3],r[2::3]
def fill4(m) : r = gis(4*m); return r[0::4],r[1::4],r[2::4],r[3::4]
MOD = 998244353
##MOD = 1000000007

def main() :
    if len(sys.argv) > 1 : global infile; infile = open(sys.argv[1],'rt')
    ## PROGRAM STARTS HERE
    T = gi()
    for tt in range(1,T+1) :
        L = gi()
        raw = " ".join([infile.readline().rstrip() for _ in range(L)] )
        tokens = raw.replace("("," ( ").replace(")"," ) ").rstrip().lstrip().split()
        Treenode = namedtuple("Treenode","w f t1 t0")
        tree = []
        def tparse(curs) :
            tidx = len(tree); tree.append(Treenode(float(tokens[curs+1]),"",-1,-1))
            if tokens[curs+2] == ")" : return curs+2
            w = tree[tidx].w; f = tokens[curs+2]; t1 = tidx+1
            curs = tparse(curs+3); t0 = len(tree); curs = tparse(curs+1)
            tree[tidx] = Treenode(w,f,t1,t0)
            return curs+1
        tparse(0)
        features = defaultdict(bool)
        def evaltree(tidx) :
            if tree[tidx].f == "" : return tree[tidx].w
            if features[tree[tidx].f] : return tree[tidx].w * evaltree(tree[tidx].t1)
            return tree[tidx].w * evaltree(tree[tidx].t0)
        print(f"Case #{tt}:")
        A = gi()
        for i in range(A) :
            aname = gs(); n = gi(); farr = [gs() for _ in range(n)]
            for f in farr : features[f] = True
            ans = evaltree(0)
            print(ans)
            for f in farr : features[f] = False

if __name__ == "__main__" :
    main()

File: codejam/cpp/2009/1B/1B_DecisionTree.cpp
#include <bits/stdc++.h>
using namespace std;
typedef long long ll;
typedef vector<ll> vi;
typedef pair<ll,ll> pi;
#define FOR(i,a) for (ll i = 0; i < (a); i++)
#define len(x) (ll) x.size()
const ll INF = 1LL << 62;
const ll MOD = 1000000007;
//const ll MOD = 998244353;
const double PI = 4*atan(double(1.0));
vector<vector<ll>> twodi(ll N, ll M, ll v) { vector<vector<ll>> res(N); for (auto &vv : res) vv.resize(M,v); return res; }

struct treenode { double w; string f; ll t1; ll t0; };

void addTokens(string &buf, vector<string> &tokens) {
    size_t p = 0; auto n = buf.size();
    while (p < n) {
        while (p < n && buf[p] == ' ') p++; if (p == n) break;
        if (buf[p] == '(') { tokens.push_back("("); p++; continue; }
        if (buf[p] == ')') { tokens.push_back(")"); p++; continue; }
        auto st = p; while (p < n && buf[p] != '(' && buf[p] != ')' && buf[p] != ' ') p++;
        tokens.push_back(buf.substr(st,p-st));
    }
}
vector<string> splitString(string s) {
    size_t p = 0; auto n = s.size(); vector<string> res; 
    while (p < n) {
        while (p < n && s[p] == ' ') p++; if (p == n) break;
        auto st = p; while (p < n && s[p] != ' ') p++;
        res.push_back(s.substr(st,p-st));
    }
    return res;
}

int main (int argc, char **argv) {
    if (argc > 1) { freopen(argv[1],"r",stdin); }
    ios_base::sync_with_stdio(false);cin.tie(0);
    // PROGRAM STARTS HERE
    string buf;
    getline(cin,buf); ll T = stoll(buf);
    for (ll tt=1;tt<=T;tt++) {
        getline(cin,buf); ll L = stoll(buf);
        vector<string> tokens;
        FOR(i,L) { getline(cin,buf); addTokens(buf,tokens); }
        fflush(stdout);
        vector<treenode> tree;
        function<ll(ll)> parsetree;
        parsetree = [&](ll idx) -> ll {
            auto tidx = (ll) tree.size();
            tree.push_back({1.0,"",-1,-1});
            tree[tidx].w = stod(tokens[idx+1]);
            if (tokens[idx+2] == ")") { return idx+2; }
            tree[tidx].f = tokens[idx+2];
            tree[tidx].t1 = (ll) tree.size();
            auto last1 = parsetree(idx+3);
            tree[tidx].t0 = (ll) tree.size();
            auto last2 = parsetree(last1+1);
            return last2+1;
        };
        parsetree(0);
        function<double(ll)> evaltree;
        set<string> features;
        evaltree = [&](ll tidx) -> double {
            auto tt = tree[tidx];
            auto ans = tt.w;
            ll c = features.count(tt.f);
            return ans * (tt.f == "" ? 1.0 : c > 0 ? evaltree(tt.t1) : evaltree(tt.t0));
        };
        getline(cin,buf); ll A = stoll(buf);
        printf("Case #%lld:\n",tt);
        FOR(i,A) {
            getline(cin,buf); auto sarr = splitString(buf); ll n = len(sarr);
            features.clear(); for (ll i = 2; i < n; i++) { features.insert(sarr[i]); }
            printf("%.17g\n",evaltree(0));
        }
    }
}

